File: mParticle-Apple-SDK/Utils/MPLaunchInfo.m
#import "MPLaunchInfo.h"
#import "MPIConstants.h"
#import <UIKit/UIKit.h>
#import "MParticleSwift.h"
#import "MPILogger.h"
#import "mParticle.h"

@interface MPLaunchInfo() {
    NSString *sourceApp;
}

@end


@implementation MPLaunchInfo

- (instancetype)initWithURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation {
    self = [super init];
    if (!self || MPIsNull(url)) {
        return nil;
    }
    
    sourceApp = sourceApplication;
    self.url = url;
    self.annotation = annotation;
    
    NSMutableDictionary *options = [[NSMutableDictionary alloc] initWithCapacity:2];
    
    if (_sourceApplication) {
        options[UIApplicationOpenURLOptionsSourceApplicationKey] = _sourceApplication;
    }
    
    if (_annotation) {
        options[UIApplicationOpenURLOptionsAnnotationKey] = _annotation;
    }
    
    if (options.count > 0) {
        _options = [options copy];
    }
    
    return self;
}

- (nonnull instancetype)initWithURL:(nonnull NSURL *)url options:(nullable NSDictionary<NSString *, id> *)options {
    self = [super init];
    if (!self || MPIsNull(url)) {
        return nil;
    }
    
    _options = options;
    sourceApp = options[UIApplicationOpenURLOptionsSourceApplicationKey];
    self.annotation = options[UIApplicationOpenURLOptionsAnnotationKey];
    self.url = url;
    
    return self;
}

#pragma mark Public accessors
- (void)setAnnotation:(id)annotation {
    Class NSDateClass = [NSDate class];
    Class NSDataClass = [NSData class];
    Class NSDictionaryClass = [NSDictionary class];
    Class NSArrayClass = [NSArray class];
    Class NSSetClass = [NSSet class];
    
    id (^valueFromObject)(id) = ^(id obj) {
        id value = nil;
        
        if ([obj isKindOfClass:NSDateClass]) {
            value = [MPDateFormatter stringFromDateRFC3339:obj];
        } else if ([obj isKindOfClass:NSDataClass] || [obj isKindOfClass:NSDictionaryClass] || [obj isKindOfClass:NSArrayClass] || [obj isKindOfClass:NSSetClass]) {
            value = nil;
        } else {
            value = obj;
        }
        
        return value;
    };
    
    NSString * (^serializeDataObject)(id) = ^(id annotationObject) {
        NSData *annotationData = nil;
        NSError *error = nil;
        
        @try {
            annotationData = [NSJSONSerialization dataWithJSONObject:annotationObject options:0 error:&error];
        } @catch (NSException *exception) {
            MPILogError(@"Error serializing annotation from app launch: %@", annotationObject);
            return (NSString *)nil;
        }
        
        NSString *serializedAnnotation = !error ? [[NSString alloc] initWithData:annotationData encoding:NSUTF8StringEncoding] : nil;
        return serializedAnnotation;
    };
    
    if ([annotation isKindOfClass:[NSDictionary class]]) {
        __block NSMutableDictionary *annotationDictionary = [[NSMutableDictionary alloc] initWithCapacity:((NSDictionary *)annotation).count];
        
        [((NSDictionary *)annotation) enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull key, id _Nonnull obj, BOOL * _Nonnull stop) {
            id value = valueFromObject(obj);
            
            if (value) {
                annotationDictionary[key] = value;
            }
        }];
        
        _annotation = serializeDataObject(annotationDictionary);
    } else if ([annotation isKindOfClass:[NSArray class]]) {
        __block NSMutableArray *annotationArray = [[NSMutableArray alloc] initWithCapacity:((NSArray *)annotation).count];
        
        [((NSArray *)annotation) enumerateObjectsUsingBlock:^(id _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
            id value = valueFromObject(obj);
            
            if (value) {
                [annotationArray addObject:value];
            }
        }];
        
        _annotation = serializeDataObject(annotationArray);
    } else if ([annotation isKindOfClass:[NSString class]]) {
        _annotation = annotation;
    } else if ([annotation isKindOfClass:[NSNumber class]]) {
        _annotation = [annotation stringValue];
    } else if ([annotation isKindOfClass:NSDateClass]) {
        _annotation = valueFromObject(annotation);
    } else {
        _annotation = nil;
    }
}

- (void)setUrl:(NSURL *)url {
    _url = url;
    
    if (sourceApp) {
        NSString *urlString = [url absoluteString];
        NSRange appLinksRange = [urlString rangeOfString:@"al_applink_data"];
        _sourceApplication = appLinksRange.location == NSNotFound ? sourceApp : [[NSString alloc] initWithFormat:@"AppLink(%@)", sourceApp];
    } else {
        _sourceApplication = nil;
    }
}

@end

File: mParticle-Apple-SDK/Utils/MPLaunchInfo.swift
//
//  MPLaunchInfo.swift
//  mParticle-Apple-SDK
//
//  Created by Brandon Stalnaker on 2/11/25.
//

import Foundation
@objc public class MPLaunchInfo : NSObject {
    private let annotationKey = UIApplication.OpenURLOptionsKey.annotation
    private let sourceAppKey = UIApplication.OpenURLOptionsKey.sourceApplication
    private var sourceApp: String?
    
    @objc public private(set) var sourceApplication: String?
    @objc public private(set) var annotation: String?
    @objc public private(set) var url: URL
    
    @objc required public init(URL: URL, sourceApplication: String?, annotation: Any?) {
        sourceApp = sourceApplication
        self.url = URL
        if let sourceApp = sourceApp {
            let urlString = url.absoluteString
            let appLinksRange = urlString.range(of: "al_applink_data")
            self.sourceApplication = appLinksRange?.lowerBound == nil ? sourceApp : String(format: "AppLink(%@)", sourceApp)
        } else {
            self.sourceApplication = nil
        }
        self.annotation = MPLaunchInfo.stringifyAnnotation(annotation)
        
        super.init()
    }
    
    @objc public init(URL: URL, options: [String : Any]?) {
        self.url = URL
        if let options = options {
            if let sourceApp = options[self.sourceAppKey.rawValue] as? String {
                self.sourceApp = sourceApp
            }
            if let sourceApp = self.sourceApp {
                let urlString = url.absoluteString
                let appLinksRange = urlString.range(of: "al_applink_data")
                self.sourceApplication = appLinksRange?.lowerBound == nil ? sourceApp : String(format: "AppLink(%@)", sourceApp)
            } else {
                self.sourceApplication = nil
            }
            
            if let annotation = options[self.annotationKey.rawValue] as? String {
                self.annotation = annotation
            }
        }
        
        super.init()
    }

    private class func stringifyAnnotation(_ annotation: Any?) -> String? {
        if let stringAnnotation = annotation as? String {
            return stringAnnotation
        }
        if let numberAnnotation = annotation as? NSNumber {
            return numberAnnotation.stringValue
        }
        if let dateAnnotation = annotation as? Date {
            return MPDateFormatter.string(fromDateRFC3339: dateAnnotation)
        }
        if let dictionaryAnnotation = annotation as? [String: Any?] {
            var jsonData: Data?
            var stringDict: [String: String] = [:]
            for key in dictionaryAnnotation.keys {
                if let value = dictionaryAnnotation[key] {
                    stringDict[key] = stringifyAnnotation(value)
                }
            }
            do {
                jsonData = try JSONSerialization.data(withJSONObject: stringDict, options: [])
            } catch {
                MPLog.error("Error serializing annotation from app launch: \(dictionaryAnnotation)")
            }
            if let jsonData = jsonData {
                return String(data: jsonData, encoding: .utf8)
            } else {
                return nil
            }
        }
        if let arrayAnnotation = annotation as? [Any] {
            var jsonData: Data?
            var stringArray: [String] = []
            for value in arrayAnnotation {
                if let value = stringifyAnnotation(value) {
                    stringArray.append(value)
                }
            }
            do {
                jsonData = try JSONSerialization.data(withJSONObject: stringArray, options: [])
            } catch {
                MPLog.error("Error serializing annotation from app launch: \(arrayAnnotation)")
            }
            if let jsonData = jsonData {
                return String(data: jsonData, encoding: .utf8)
            } else {
                return nil
            }
        } else {
            return nil
        }
    }
    
    @objc public var options: [String : Any] {
        get {
            var options: [String : Any] = [:]
            
            if let sourceApplication = self.sourceApplication {
                options[self.sourceAppKey.rawValue] = sourceApplication
            }
            
            if let annotation = self.annotation {
                options[self.annotationKey.rawValue] = annotation
            }
            
            return options
        }
    }
}
